src/psyclone/psyir/nodes/array_mixin.py

```diff
@@ -120,12 +121,73 @@ def _validate_index(self, index):
                 f"'{index}' must be less than the number of dimensions "
                 f"'{len(self.indices)}'.")
 
+    def _is_bound_op(self, expr, bound_operator, index):
+        '''Utility function that checks that the provided 'expr' argument is
+        in the form '[UL]BOUND(array_name, index)', where the type of
+        bound operation is determined by the 'bound_operator'
+        argument, array_name is the name of this array and the 'index'
+        argument provides the index value.
+
+        :param expr: a PSyIR expression.
+        :type expr: :py:class:`psyclone.psyir.nodes.Node`
+        :param bound_operator: the particular bound operation.
+        :type bound_operator: \
+            :py:class:`psyclone.psyir.nodes.operation.BinaryOperation.\
+            Operator.LBOUND` or :py:class:`psyclone.psyir.nodes.operation.\
+            BinaryOperation.Operator.UBOUND`
+        :param int index: the bounds index.
+
+        :returns: True if the expr is in the expected form and False \
+            otherwise.
+        :rtype: bool
+
+        '''
+        if (isinstance(expr, BinaryOperation) and
+                expr.operator == bound_operator):
+            # This is the expected bound
+            if self.is_same_array(expr.children[0]):
+                # The arrays match
+                if (isinstance(expr.children[1], Literal) and
+                        expr.children[1].datatype.intrinsic ==
+                        ScalarType.Intrinsic.INTEGER
+                        and expr.children[1].value == str(index+1)):
+                    # This is the correct index
+                    return True
+        return False
+
+    def _get_symbol_declaration(self):
```

Comment:
Can't you just use the `datatype` method?

Reply:
No, as `ArrayMember` subclasses  `ArrayMixin` and does not have a datatype method. I have to say I'm not quite sure how you get to the type of an ArrayMember (or Member for that matter), looking at its implementation.

Reply:
Also, if the type is UnkownFortranType we don't get an ArrayType which is what we need.

Reply:
The implementation of `lbound` and `ubound` was a mess, as you noticed. I've re-written. Hopefully things will be clearer in the updated version.

Reply:
They are clearer thanks but it still looks to me as though `self.datatype` might do the job. Please could you investigate and explain if necessary :-)

Reply:
I think this should be set to resolved as this method no longer exists.